### PR TITLE
Enhanced Android TV Support + DPAD navigation

### PR DESCRIPTION
**Pull request by nikiink**
Add enhanced Android TV support:

- On some Android TV as the WallpaperManager is missing the app crashes at startup (now it does not crash)
- On first startup if on Android TV the desktop orientation is set to landscape as default
- On some Android TV as the WallpaperManager is missing, is not possible to set a system wallpaper, so i added an Android TV setting where to user can specify a wallpaper that set the wallpaper without using WallpaperManager.
- Minimal DPAD navigation

### CODE REVIEW

**Comment by gsantner** [comment]
How stable is this? Do you have it in use for some time?

I don't have any real device Android TV to test.

**Comment by tonio-nucci** [comment]
> 
> 
> How stable is this? Do you have it in use for some time?
> 
> I don't have any real device Android TV to test.

Yes, i use it on Android T9 TV box.

File: app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java [hunk 1]
```diff
@@ -222,6 +227,16 @@ private void init() {
...
             Setup.appSettings().setAppFirstLaunch(false);
+
+            // tonio-nucci - Check at first run if on android TV (go landscape)
+            UiModeManager uiModeManager = (UiModeManager) getSystemService(UI_MODE_SERVICE);
+            if (uiModeManager.getCurrentModeType() == Configuration.UI_MODE_TYPE_TELEVISION) {
+                Log.i(this.getClass().getName(), "First start - Running on a TV Device - Setting landscape");
+                Setup.appSettings().get().setString(R.string.pref_key__desktop_orientation, Definitions.DESKTOP_ORIENTATION_LANDSCAPE);
+            } else {
+                Log.i(this.getClass().getName(), "First start - Running on a non-TV Device");
+            }
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:238]: remove log

File: app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java [hunk 2]
```diff
@@ -594,6 +609,41 @@ private void handleLauncherResume() {
...
+                if (wallpaperPath!=null) {
+                    getBackground().setBackground(Drawable.createFromPath(wallpaperPath));
+                    getBackground().setVisibility(View.VISIBLE);
+                }
+            } else {
+                //remove background
+                getBackground().setBackgroundResource(0);
+                getBackground().setVisibility(View.INVISIBLE);
+                //getBackground().setBackgroundColor(Color.RED);
+            }
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:628]: remove log, remove debug code where there is still one

File: app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java [hunk 3]
```diff
@@ -594,6 +609,41 @@ private void handleLauncherResume() {
...
+
+            // tonio-nucci (for DPAD/TV)
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:630]: remove your name  (tonio-nucci) from comments please

File: app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java [hunk 4]
```diff
@@ -594,6 +609,41 @@ private void handleLauncherResume() {
...
+            getDesktop().addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
+                 @Override
+                 public void onPageScrolled(int i, float v, int i1) { }
+
+                 @Override
+                 public void onPageSelected(int i) {
+                     Log.i(this.getClass().getName(), "Page selected: " + i);
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/HomeActivity.java:637]: log

File: app/src/main/java/com/benny/openlauncher/activity/SettingsActivity.java [hunk 5]
```diff
@@ -77,7 +79,15 @@ protected void onActivityResult(int requestCode, int resultCode, final Intent da
                     BackupHelper.restoreConfig(this, Utils.getFileForUri(files.get(0)).toString());
                     System.exit(0);
                     break;
+                case Definitions.ANDROID_TV_PICK_WALLPAPER:
+                    Log.i(this.getClass().getName(), "Android Tv Picked Wallpaper");
+                    String imagePath = Utils.getFileForUri(files.get(0)).toString();
+                    Log.i(this.getClass().getName(), "Wallpaper Path: " + imagePath);
+                    Setup.dataManager().open(); // without this, you get sqlite db already closed error
+                    _appSettings.setAndroidTvWallpaper(imagePath);
+                    break;
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/activity/SettingsActivity.java:88]: log

File: app/src/main/java/com/benny/openlauncher/util/AppSettings.java [hunk 6]
```diff
@@ -339,6 +339,13 @@ public void setDesktopLock(boolean value) {
         setBool(R.string.pref_key__desktop_lock, value);
     }
 
+    // Android TV
+    public String getAndroidTvWallpaper() { return getString(R.string.pref_key__android_tv_settings_wallpaper, null); }
+    public void setAndroidTvWallpaper(String value) { setString(R.string.pref_key__android_tv_settings_wallpaper, value); }
+
+    public boolean getAndroidTvOverrideSystemWallpaper() { return getBool(R.string.pref_key__android_tv_settings_override_system_wallpaper, false); }
+    ///
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/util/AppSettings.java:347]: always {

}

File: app/src/main/java/com/benny/openlauncher/util/Definitions.java [hunk 7]
```diff
@@ -27,4 +31,8 @@
         Inverse,
         Off
     }
+
+    public static final String DESKTOP_ORIENTATION_PORTRAIT  = "0";
+    public static final String DESKTOP_ORIENTATION_SENSOR    = "1";
+    public static final String DESKTOP_ORIENTATION_LANDSCAPE = "2";
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/util/Definitions.java:37]: unused?

File: app/src/main/java/com/benny/openlauncher/widget/AppDrawerGrid.java [hunk 8]
```diff
@@ -109,7 +109,18 @@ public void onClick(View v) {
                             Tool.startApp(v.getContext(), app, null);
                         }
                     })
-                    .withOnLongClickListener(DragHandler.getLongClick(Item.newAppItem(app), DragAction.Action.DRAWER, null)));
+                    .withOnLongClickListener(DragHandler.getLongClick(Item.newAppItem(app), DragAction.Action.DRAWER, null))
+                    .withOnFocusChangeListener(new OnFocusChangeListener() { // tonio-nucci (for DPAD/TV)
+                            @Override
+                            public void onFocusChange(View v, boolean hasFocus) {
+                                if (hasFocus) {
+                                    v.setBackgroundColor(Color.argb(128, 255,255,255));
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/widget/AppDrawerGrid.java:117]: migrate to color resource, or reuse a existing java Color. or R.color. if available.

File: app/src/main/java/com/benny/openlauncher/widget/AppItemView.java [hunk 9]
```diff
@@ -190,6 +190,9 @@ public AppItemView getView() {
         }
 
         public Builder setAppItem(final Item item) {
+            //tonio-nucci (for DPAD/TV)
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/widget/AppItemView.java:193]: tonio-nucci

File: app/src/main/java/com/benny/openlauncher/widget/GroupPopupView.java [hunk 10]
```diff
@@ -110,6 +110,9 @@ public boolean showPopup(final Item item, final View itemView, final DesktopCall
 
         boolean appsChanged = false;
 
+        //tonio-nucci (for DPAD/TV) - focus on the first item
+        boolean focusSet = false;
```
Comment (gsantner) [inline app/src/main/java/com/benny/openlauncher/widget/GroupPopupView.java:114]: maybe rename the var so it contains androidTV? as it's completley tv related

File: app/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -70,6 +70,8 @@
     <string name="pref_title__dock">Dock</string>
     <string name="pref_title__app_drawer">App Drawer</string>
     <string name="pref_title__appearance">Appearance</string>
+    <string name="pref_title__android_tv">Android TV</string>
+    <string name="pref_title__android_tv_wallpaper">Wallpaper</string>
```
Comment (gsantner) [inline app/src/main/res/values/strings.xml:74]: Can be both in untranslatable

**Review by gsantner — CHANGES_REQUESTED** [review]
@tonio-nucci
I looked through the code (late but still). Here is my review.

When code quality gets OK, and you can assure it works as expected on TV - I plan to merge it.

**Comment by gsantner** [comment]
Without reaction I will close this stale pull request in 30 days of inactivity.

**Comment by gsantner** [comment]
closing due inactivity

**Pull request closed without merging**